File: voxtral_realtime/macos/VoxtralRealtime/Models/Session.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Foundation

enum SessionSource: String, Codable, Sendable, Hashable {
    case transcription
    case dictation
}

struct Session: Identifiable, Codable, Sendable, Hashable {
    let id: UUID
    let date: Date
    var title: String
    var transcript: String
    var duration: TimeInterval
    var source: SessionSource
    var rawTranscript: String?
    var tags: [String]
    var wakeTriggered: Bool
    var pinned: Bool
    var usedSnippetIDs: [UUID]

    init(
        id: UUID = UUID(),
        date: Date = .now,
        title: String = "",
        transcript: String = "",
        duration: TimeInterval = 0,
        source: SessionSource = .transcription,
        rawTranscript: String? = nil,
        tags: [String] = [],
        wakeTriggered: Bool = false,
        pinned: Bool = false,
        usedSnippetIDs: [UUID] = []
    ) {
        self.id = id
        self.date = date
        self.title = title
        self.transcript = transcript
        self.duration = duration
        self.source = source
        self.rawTranscript = rawTranscript
        self.tags = tags
        self.wakeTriggered = wakeTriggered
        self.pinned = pinned
        self.usedSnippetIDs = usedSnippetIDs
    }

    enum CodingKeys: String, CodingKey {
        case id
        case date
        case title
        case transcript
        case duration
        case source
        case rawTranscript
        case tags
        case wakeTriggered
        case pinned
        case usedSnippetIDs
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decodeIfPresent(UUID.self, forKey: .id) ?? UUID()
        date = try container.decodeIfPresent(Date.self, forKey: .date) ?? .now
        title = try container.decodeIfPresent(String.self, forKey: .title) ?? ""
        transcript = try container.decodeIfPresent(String.self, forKey: .transcript) ?? ""
        duration = try container.decodeIfPresent(TimeInterval.self, forKey: .duration) ?? 0
        source = try container.decodeIfPresent(SessionSource.self, forKey: .source) ?? .transcription
        rawTranscript = try container.decodeIfPresent(String.self, forKey: .rawTranscript)
        tags = try container.decodeIfPresent([String].self, forKey: .tags) ?? []
        wakeTriggered = try container.decodeIfPresent(Bool.self, forKey: .wakeTriggered) ?? false
        pinned = try container.decodeIfPresent(Bool.self, forKey: .pinned) ?? false
        usedSnippetIDs = try container.decodeIfPresent([UUID].self, forKey: .usedSnippetIDs) ?? []
    }

    var displayTitle: String {
        if !title.isEmpty { return title }
        let formatter = DateFormatter()
        formatter.dateStyle = .medium
        formatter.timeStyle = .short
        return formatter.string(from: date)
    }

    var previewText: String {
        transcript.isEmpty ? rawTranscript ?? "" : transcript
    }
}

File: voxtral_realtime/windows/VoxtralRealtime/VoxtralRealtime/Models/Session.cs
// Copyright (c) Meta Platforms, Inc. and affiliates.
// All rights reserved.
// This source code is licensed under the BSD-style license found in the
// LICENSE file in the root directory of this source tree.

using System.Text.Json.Serialization;

namespace VoxtralRealtime.Models;

public class Session
{
    [JsonPropertyName("id")]
    public Guid Id { get; set; } = Guid.NewGuid();

    [JsonPropertyName("date")]
    public DateTime Date { get; set; } = DateTime.Now;

    [JsonPropertyName("title")]
    public string? Title { get; set; }

    [JsonPropertyName("transcript")]
    public string Transcript { get; set; } = "";

    [JsonPropertyName("duration")]
    public double Duration { get; set; }

    [JsonPropertyName("source")]
    [JsonConverter(typeof(JsonStringEnumConverter))]
    public SessionSource Source { get; set; } = SessionSource.Transcription;

    [JsonPropertyName("rawTranscript")]
    public string? RawTranscript { get; set; }

    [JsonPropertyName("tags")]
    public List<string> Tags { get; set; } = new();

    [JsonPropertyName("wakeTriggered")]
    public bool WakeTriggered { get; set; }

    [JsonPropertyName("pinned")]
    public bool Pinned { get; set; }

    [JsonPropertyName("usedSnippetIds")]
    public List<Guid> UsedSnippetIds { get; set; } = new();

    [JsonIgnore]
    public string DisplayTitle => !string.IsNullOrEmpty(Title)
        ? Title
        : Date.ToString("MMM d, yyyy h:mm tt");

    [JsonIgnore]
    public string PreviewText
    {
        get
        {
            var text = Transcript.Trim();
            return text.Length > 100 ? text[..100] + "..." : text;
        }
    }
}
